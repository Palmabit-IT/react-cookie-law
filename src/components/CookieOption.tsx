import React from 'react';

interface Props {
  checked?: boolean
  onChange?: (checked: boolean) => void
  id?: string
  text?: string
  styles?: Record<string, React.CSSProperties>
  disabled?: boolean
}

interface State {
  checked: boolean
}

class CookieOption extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const { checked = false } = props;
    this.state = { checked };
    this.handleOnChange = this.handleOnChange.bind(this);
  }

  handleOnChange() {
    const { onChange = () => undefined } = this.props;
    const { checked } = this.state;
    const newValue = !checked;

    this.setState({ checked: newValue });
    onChange(newValue);
  }

  render() {
    const {
      id = '',
      text = '',
      styles = {},
      disabled = false,
    } = this.props;

    const isDisabled = disabled ? { disabled } : {};
    const { checked } = this.state;

    const { optionWrapperStyle, optionLabelStyle, checkboxStyle } = styles;

    return (
      <div
        className="react-cookie-law-option-wrapper"
        style={optionWrapperStyle}
      >
        <input
          type="checkbox"
          id={id}
          className="react-cookie-law-option-checkbox"
          style={checkboxStyle}
          checked={checked}
          onChange={this.handleOnChange}
          {...isDisabled}
        />
        <label htmlFor={id} style={optionLabelStyle}>
          {text}
        </label>
      </div>
    );
  }
}

export default CookieOption;
