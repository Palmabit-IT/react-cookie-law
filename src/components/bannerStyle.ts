const bannerStyle = {
  dialog: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    zIndex: 100000,
    backgroundColor: '#f8f7f7',
    padding: '10px',
  },
  container: {
    maxWidth: '960px',
    marginLeft: 'auto',
    marginRight: 'auto',
    overflow: 'hidden',
  },
  message: {
    minHeight: '32px',
    fontSize: '10pt',
    fontWeight: 400,
    lineHeight: '130%',
    padding: '10px 0',
    color: '#000000',
  },
  policy: {
    fontSize: '10pt',
    marginLeft: '10px',
    color: '#000000',
    textDecoration: 'underline',
  },
  selectPane: {
    display: 'table-cell',
    padding: '3px 0px 10px 0px',
  },
  optionWrapper: {
    display: 'inline-block',
    position: 'relative',
    marginRight: '7px',
    marginLeft: '7px',
    float: 'left',
  },
  optionLabel: {
    height: 'auto',
    width: 'auto',
    minHeight: '14px',
    fontSize: '12pt',
    color: '#c6c6c6',
    display: 'inline-block',
    padding: '1px 0 0 20px',
    position: 'relative',
    top: 0,
    left: 0,
    zIndex: 1,
    cursor: 'default',
    verticalAlign: 'top',
  },
  checkbox: {
    position: 'absolute',
    top: '4px',
    left: 0,
    width: '14px',
    height: '14px',
    zIndex: 2,
    cursor: 'pointer',
  },
  buttonWrapper: {
    float: 'right',
  },
  button: {
    display: 'inline-block',
    backgroundColor: '#000000',
    padding: '3px',
    minWidth: '80px',
    color: '#ffffff',
    textDecoration: 'none',
    fontSize: '10pt',
    fontWeight: 400,
    marginRight: '5px',
    marginLeft: '5px',
    textAlign: 'center',
    whiteSpace: 'nowrap',
    cursor: 'pointer',
  },
} as const;

export type StyleableComponents = keyof typeof bannerStyle

export default bannerStyle
