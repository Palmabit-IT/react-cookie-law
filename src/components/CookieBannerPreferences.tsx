import React from 'react';
import CookieOption from './CookieOption';
import bannerStyle, { StyleableComponents } from './bannerStyle';

import type { CoryphaPreference } from '../coryphaAPI';

export interface Props {
  styles?: Partial<Record<StyleableComponents, React.CSSProperties>>
  necessaryOptionText?: string
  preferencesOptionText?: string
  statisticsOptionText?: string
  marketingOptionText?: string
  showPreferencesOption?: boolean
  showStatisticsOption?: boolean
  showMarketingOption?: boolean
  preferencesDefaultChecked?: boolean
  statisticsDefaultChecked?: boolean
  marketingDefaultChecked?: boolean
  onTogglePreferencesCookies?: (checked?: boolean) => void
  onToggleStatisticsCookies?: (checked?: boolean) => void
  onToggleMarketingCookies?: (checked?: boolean) => void
  coryphaPreferences?: CoryphaPreference[]
  onToggleCoryphaPreference?: (active: boolean, preference: CoryphaPreference) => void
}
const CookieBannerPreferences: React.FC<Props> = (props: Props = {}) => {
  const {
    styles = {},
    necessaryOptionText = 'Necessary',
    preferencesOptionText = 'Preferences',
    statisticsOptionText = 'Statistics',
    marketingOptionText = 'Marketing',
    showPreferencesOption = true,
    showStatisticsOption = true,
    showMarketingOption = true,
    preferencesDefaultChecked = false,
    statisticsDefaultChecked = false,
    marketingDefaultChecked = false,
    onTogglePreferencesCookies = () => undefined,
    onToggleStatisticsCookies = () => undefined,
    onToggleMarketingCookies = () => undefined,
    coryphaPreferences = [],
    onToggleCoryphaPreference = () => undefined,
  } = props;

  const {
    selectPane: selectPaneStyle,
    optionWrapper: optionWrapperStyle,
    optionLabel: optionLabelStyle,
    checkbox: checkboxStyle,
  } = { ...bannerStyle, ...styles };

  const cookieOptionStyle = {
    optionWrapperStyle,
    optionLabelStyle,
    checkboxStyle,
  };

  return (
    <div className="react-cookie-law-select-pane" style={selectPaneStyle}>
      {coryphaPreferences.length > 0 ? (
        <>
          <CookieOption
            id="check-required-cookies"
            text={necessaryOptionText}
            styles={cookieOptionStyle}
            disabled
            checked
          />

          {coryphaPreferences.map((preference) => (
            <CookieOption
              key={preference.id}
              id={preference.id}
              text={preference.name}
              styles={cookieOptionStyle}
              disabled={preference.required}
              checked={preference.required}
              onChange={(value) => onToggleCoryphaPreference(value, preference)}
            />
          ))}
        </>
      ) : (
        <>
          <CookieOption
            id="check-required-cookies"
            text={necessaryOptionText}
            styles={cookieOptionStyle}
            disabled
            checked
          />

          {showPreferencesOption && (
            <CookieOption
              id="check-preferences-cookies"
              text={preferencesOptionText}
              styles={cookieOptionStyle}
              checked={preferencesDefaultChecked}
              onChange={onTogglePreferencesCookies}
            />
          )}

          {showStatisticsOption && (
            <CookieOption
              id="check-statistics-cookies"
              text={statisticsOptionText}
              styles={cookieOptionStyle}
              checked={statisticsDefaultChecked}
              onChange={onToggleStatisticsCookies}
            />
          )}

          {showMarketingOption && (
            <CookieOption
              id="check-marketing-cookies"
              text={marketingOptionText}
              styles={cookieOptionStyle}
              checked={marketingDefaultChecked}
              onChange={onToggleMarketingCookies}
            />
          )}
        </>
      )}
    </div>
  );
};

export default CookieBannerPreferences
