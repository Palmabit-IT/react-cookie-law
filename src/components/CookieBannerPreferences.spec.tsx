import React from 'react';
import { mount, shallow } from 'enzyme';
import CookieBannerPreferences from './CookieBannerPreferences';
import CookieOption from './CookieOption';
import bannerStyle from './bannerStyle';

describe('CookieBannerPreferences component', () => {
  test('should be rendered', () => {
    const component = shallow(
      <CookieBannerPreferences />,
    );

    expect(component).toMatchSnapshot();
  });

  test('should be rendered with elements', () => {
    const props = {
      styles: {},
      necessaryOptionText: 'Necessary',
      preferencesOptionText: 'Preferences',
      statisticsOptionText: 'Statistics',
      marketingOptionText: 'Marketing',
      showPreferencesOption: true,
      showStatisticsOption: true,
      showMarketingOption: true,
      preferencesDefaultChecked: false,
      statisticsDefaultChecked: false,
      marketingDefaultChecked: false,
      onTogglePreferencesCookies: () => undefined,
      onToggleStatisticsCookies: () => undefined,
      onToggleMarketingCookies: () => undefined,
    };

    const component = mount(
      <CookieBannerPreferences {...props} />,
    );

    const cookieOptionStyle = {
      optionWrapperStyle: bannerStyle.optionWrapper,
      optionLabelStyle: bannerStyle.optionLabel,
      checkboxStyle: bannerStyle.checkbox,
    };

    expect(component.contains(<CookieOption id="check-required-cookies" text="Necessary" disabled checked styles={cookieOptionStyle} />)).toBeTruthy();
    expect(component.contains(<CookieOption id="check-preferences-cookies" text="Preferences" checked={false} onChange={props.onTogglePreferencesCookies} styles={cookieOptionStyle} />)).toBeTruthy();
    expect(component.contains(<CookieOption id="check-statistics-cookies" text="Statistics" checked={false} onChange={props.onToggleStatisticsCookies} styles={cookieOptionStyle} />)).toBeTruthy();
    expect(component.contains(<CookieOption id="check-marketing-cookies" text="Marketing" checked={false} onChange={props.onToggleMarketingCookies} styles={cookieOptionStyle} />)).toBeTruthy();
  });

  test('should hide preferences checkbox', () => {
    const props = {
      onDecline: jest.fn(),
      showPreferencesOption: false,
    };

    const component = mount(
      <CookieBannerPreferences {...props} />,
    );

    expect(component.find('#check-preferences-cookies').exists()).toBeFalsy();
  });

  test('should hide statistics checkbox', () => {
    const props = {
      onDecline: jest.fn(),
      showStatisticsOption: false,
    };

    const component = mount(
      <CookieBannerPreferences {...props} />,
    );

    expect(component.find('#check-statistics-cookies').exists()).toBeFalsy();
  });

  test('should hide marketing checkbox', () => {
    const props = {
      onDecline: jest.fn(),
      showMarketingOption: false,
    };

    const component = mount(
      <CookieBannerPreferences {...props} />,
    );

    expect(component.find('#check-marketing-cookies').exists()).toBeFalsy();
  });

  test('should show all checkboxes as default', () => {
    const component = mount(
      <CookieBannerPreferences />,
    );

    expect(component.find('#check-preferences-cookies').exists()).toBeTruthy();
    expect(component.find('#check-statistics-cookies').exists()).toBeTruthy();
    expect(component.find('#check-marketing-cookies').exists()).toBeTruthy();
  });
});
