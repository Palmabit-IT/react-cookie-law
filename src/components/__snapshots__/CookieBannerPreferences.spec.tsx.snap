// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieBannerPreferences component should be rendered 1`] = `
<div
  className="react-cookie-law-select-pane"
  style={
    Object {
      "display": "table-cell",
      "padding": "3px 0px 10px 0px",
    }
  }
>
  <CookieOption
    checked={true}
    disabled={true}
    id="check-required-cookies"
    styles={
      Object {
        "checkboxStyle": Object {
          "cursor": "pointer",
          "height": "14px",
          "left": 0,
          "position": "absolute",
          "top": "4px",
          "width": "14px",
          "zIndex": 2,
        },
        "optionLabelStyle": Object {
          "color": "#c6c6c6",
          "cursor": "default",
          "display": "inline-block",
          "fontSize": "12pt",
          "height": "auto",
          "left": 0,
          "minHeight": "14px",
          "padding": "1px 0 0 20px",
          "position": "relative",
          "top": 0,
          "verticalAlign": "top",
          "width": "auto",
          "zIndex": 1,
        },
        "optionWrapperStyle": Object {
          "display": "inline-block",
          "float": "left",
          "marginLeft": "7px",
          "marginRight": "7px",
          "position": "relative",
        },
      }
    }
    text="Necessary"
  />
  <CookieOption
    checked={false}
    id="check-preferences-cookies"
    onChange={[Function]}
    styles={
      Object {
        "checkboxStyle": Object {
          "cursor": "pointer",
          "height": "14px",
          "left": 0,
          "position": "absolute",
          "top": "4px",
          "width": "14px",
          "zIndex": 2,
        },
        "optionLabelStyle": Object {
          "color": "#c6c6c6",
          "cursor": "default",
          "display": "inline-block",
          "fontSize": "12pt",
          "height": "auto",
          "left": 0,
          "minHeight": "14px",
          "padding": "1px 0 0 20px",
          "position": "relative",
          "top": 0,
          "verticalAlign": "top",
          "width": "auto",
          "zIndex": 1,
        },
        "optionWrapperStyle": Object {
          "display": "inline-block",
          "float": "left",
          "marginLeft": "7px",
          "marginRight": "7px",
          "position": "relative",
        },
      }
    }
    text="Preferences"
  />
  <CookieOption
    checked={false}
    id="check-statistics-cookies"
    onChange={[Function]}
    styles={
      Object {
        "checkboxStyle": Object {
          "cursor": "pointer",
          "height": "14px",
          "left": 0,
          "position": "absolute",
          "top": "4px",
          "width": "14px",
          "zIndex": 2,
        },
        "optionLabelStyle": Object {
          "color": "#c6c6c6",
          "cursor": "default",
          "display": "inline-block",
          "fontSize": "12pt",
          "height": "auto",
          "left": 0,
          "minHeight": "14px",
          "padding": "1px 0 0 20px",
          "position": "relative",
          "top": 0,
          "verticalAlign": "top",
          "width": "auto",
          "zIndex": 1,
        },
        "optionWrapperStyle": Object {
          "display": "inline-block",
          "float": "left",
          "marginLeft": "7px",
          "marginRight": "7px",
          "position": "relative",
        },
      }
    }
    text="Statistics"
  />
  <CookieOption
    checked={false}
    id="check-marketing-cookies"
    onChange={[Function]}
    styles={
      Object {
        "checkboxStyle": Object {
          "cursor": "pointer",
          "height": "14px",
          "left": 0,
          "position": "absolute",
          "top": "4px",
          "width": "14px",
          "zIndex": 2,
        },
        "optionLabelStyle": Object {
          "color": "#c6c6c6",
          "cursor": "default",
          "display": "inline-block",
          "fontSize": "12pt",
          "height": "auto",
          "left": 0,
          "minHeight": "14px",
          "padding": "1px 0 0 20px",
          "position": "relative",
          "top": 0,
          "verticalAlign": "top",
          "width": "auto",
          "zIndex": 1,
        },
        "optionWrapperStyle": Object {
          "display": "inline-block",
          "float": "left",
          "marginLeft": "7px",
          "marginRight": "7px",
          "position": "relative",
        },
      }
    }
    text="Marketing"
  />
</div>
`;
